### PR TITLE
[WIN32SS][LPK] Add BiDi support to menus and window captions

### PR DESCRIPTION
**Pull request by peterooch**
## Purpose

Added callback function that redirects calls to GreExtTextOutW that didn't went through lpk BiDi processing, calls that are from the kernel.

Completely solves JIRA issue [CORE-6910](https://jira.reactos.org/browse/CORE-6910)

### CODE REVIEW

**Comment by peterooch** [comment]
virtualbox_reactos_10_08_2018_14_38_08

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -305,6 +305,100 @@ SharedFace_Release(PSHARED_FACE Ptr)
...
+    Argument->hdc    = hdc;
+    Argument->x      = x;
+    Argument->y      = y;    
+    Argument->flags  = flags;
+    Argument->count  = count;
+
+    /* copy rect */
+    if (lprc)
+    {
+        Argument->rect.left   = lprc->left;
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/freetype.c]: maybe Argument->rect = *lprc;

File: win32ss/gdi/ntgdi/freetype.c [hunk 2]
```diff
@@ -305,6 +305,100 @@ SharedFace_Release(PSHARED_FACE Ptr)
...
+        {
+            DPRINT1("Failed to copy result from user mode!\n");
+            Status = _SEH2_GetExceptionCode();
+        }
+        _SEH2_END;
+    }
+
+    if (!NT_SUCCESS(Status))
+    {
+        *bResult  = FALSE;
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/freetype.c]: extra space.

File: win32ss/gdi/ntgdi/freetype.c [hunk 3]
```diff
@@ -305,6 +305,100 @@ SharedFace_Release(PSHARED_FACE Ptr)
...
+        
+    UserEnterCo();
+ 
+    IntCbFreeMemory(Argument);
+        
+    if (NT_SUCCESS(Status))
+    {
+        _SEH2_TRY
+        {
+            ProbeForRead(ResultPointer, sizeof(HMODULE), 1);
```
[the remark this answers is not in the document]
Comment (peterooch) [reply win32ss/gdi/ntgdi/freetype.c]: This code is based on what is in `co_IntClientLoadLibrary`, which also a callback that uses string parameters. but will check if can be changed to BOOL

File: win32ss/user/user32/misc/dllmain.c [hunk 4]
```diff
@@ -641,3 +642,24 @@ User32CallOBMFromKernel(PVOID Arguments, ULONG ArgumentLength)
...
 }
+
+NTSTATUS WINAPI User32CallLPKFromKernel(PVOID Arguments, ULONG ArgumentLength)
+{
+    BOOL bResult;
+    PLPK_CALLBACK_ARGUMENTS Argument;
+
+    Argument = (PLPK_CALLBACK_ARGUMENTS)Arguments;
+
+    Argument->lpString = (PWCHAR)((ULONG_PTR)Argument->lpString + (ULONG_PTR)Argument);
```
[the remark this answers is not in the document]
Comment (peterooch) [reply win32ss/user/user32/misc/dllmain.c]: See `co_IntClientLoadLibrary` and `User32CallClientLoadLibraryFromKernel`

File: win32ss/gdi/ntgdi/freetype.c (hunk 1 above)
Comment (peterooch) [reply win32ss/gdi/ntgdi/freetype.c]: Thank you

**Comment by peterooch** [comment]
virtualbox_reactos_10_08_2018_14_38_08

**Comment by HBelusca** [comment]
"Grey +" ? What's that?

**Comment by peterooch** [comment]
> "Grey +" ? What's that?
A 7-zip menu thing...

**Comment by HBelusca** [comment]
Ah ok, so that's what they write instead of "Ctrl-+" etc...

File: win32ss/user/rtl/text.c [hunk 5]
```diff
@@ -939,6 +939,100 @@ static void TEXT_DrawUnderscore (HDC hdc, int x, int y, const WCHAR *str, int of
...
+    }
+
+    if (!NT_SUCCESS(Status))
+    {
+        goto fallback;
+    }
+
+    return bResult;
+
+    fallback:
```
Comment (HBelusca) [inline win32ss/user/rtl/text.c]: Please remove the indentation.

File: win32ss/user/rtl/text.c [hunk 6]
```diff
@@ -939,6 +939,100 @@ static void TEXT_DrawUnderscore (HDC hdc, int x, int y, const WCHAR *str, int of
...
+        return GreExtTextOutW(hdc, x, y, flags, lprc, lpString, count, NULL, 0);
```
Comment (HBelusca) [inline win32ss/user/rtl/text.c]: Reduce the indentation to 4 spaces.

File: win32ss/user/user32/misc/dllmain.c [hunk 7]
```diff
@@ -641,3 +642,24 @@ User32CallOBMFromKernel(PVOID Arguments, ULONG ArgumentLength)
...
+NTSTATUS WINAPI User32CallLPKFromKernel(PVOID Arguments, ULONG ArgumentLength)
+{
+    BOOL bResult;
+    PLPK_CALLBACK_ARGUMENTS Argument;
+
+    Argument = (PLPK_CALLBACK_ARGUMENTS)Arguments;
+
+    Argument->lpString = (LPWSTR)((ULONG_PTR)Argument->lpString + (ULONG_PTR)Argument);
+
+    bResult = ExtTextOutW(Argument->hdc,
```
Comment (HBelusca) [inline win32ss/user/user32/misc/dllmain.c]: In the future this certainly will have to be generalized.

Comment (peterooch) [reply win32ss/user/user32/misc/dllmain.c]: What do you mean by "generalization"?

Comment (HBelusca) [reply win32ss/user/user32/misc/dllmain.c]: Err forget what I've just said. I indeed see in the user32 symbols that there is indeed a special callback for ExtTextOutW (ClientExtTextOutW).
https://www.reactos.org/wiki/Techwiki:Win32k/apfnDispatch

**Comment by yagoulas** [comment]
Adding an extra callback is the way to go but morr changes are needed in the kernel mode part because doing a callback can have unpreidctable results. The thread coukd die  in user mode and never return to free resources or the window pwnd could be destroyed while in user mode and we could  end up with a stale pwnd pointrr this way.

**Comment by ThFabba** [comment]
Hmm I don't see any such issues looking at the code. Are you able to point to specific instances?
The memory allocation uses IntCbAllocateMemory which gets added to the thread list, so is safe for callouts, and there's no PWND to be seen, only an HDC, which should be safe as well as far as I understand.
Thanks!

File: win32ss/user/rtl/text.c [hunk 8]
```diff
@@ -939,6 +939,98 @@ static void TEXT_DrawUnderscore (HDC hdc, int x, int y, const WCHAR *str, int of
...
+                     INT y,
+                     UINT flags,
+                     PRECTL lprc,
+                     LPCWSTR lpString,
+                     UINT count)
+{
+    PVOID ResultPointer;
+    ULONG ResultLength;
+    ULONG ArgumentLength;
+    ULONG_PTR pStringBuffer = 0;
```
Comment (learn-more) [inline win32ss/user/rtl/text.c]: This does not need to be zero-initialized.

File: win32ss/user/rtl/text.c [hunk 9]
```diff
@@ -939,6 +939,98 @@ static void TEXT_DrawUnderscore (HDC hdc, int x, int y, const WCHAR *str, int of
...
+    {
+        RtlZeroMemory(&Argument->rect, sizeof(RECT));
+        Argument->bRect = FALSE;
+    }
+
+    /* Align lpString
+       mimicks code from co_IntClientLoadLibrary */
+    Argument->lpString = (LPWSTR)pStringBuffer;
+    pStringBuffer += (ULONG_PTR)Argument;
+    RtlStringCchCopyNW((LPWSTR)pStringBuffer, count + 1, lpString, count);
```
Comment (learn-more) [inline win32ss/user/rtl/text.c]: You should probably check the return from this.

**Comment by peterooch** [comment]
Any opinions on should I include the changes that are mentioned in:
https://www.reactos.org/forum/viewtopic.php?f=9&t=17378
or should I keep those for a later PR?

**Comment by HBelusca** [comment]
Hi! I think that the changes mentioned in the forum should be done in a separate PR.

**Pull request merged**